### ISSUE TITLE
Show VM based workloads under the Workloads tab

### ISSUE THREAD

**Issue by nrfox** [kiali/kiali#4169]
`WorkloadEntry` and `WorkloadGroup` workloads are not displayed ~on the graph or~ on the `workloads` tab. ~The effect of this is that any traffic destined for a VM workload is not represented within Kiali.~ Kiali should display _VM based workloads_ ~destined for a `WorkloadEntry` and/or `WorkloadGroup` both on the graph and~ in the workloads tab.

With the following `WorkloadEntries` deployed:
```
NAMESPACE   NAME             AGE   ADDRESS
bookinfo    details-v1       42m   10.10.0.212
bookinfo    productpage-v1   42m   10.10.0.213
bookinfo    ratings-v1       42m   10.10.0.215
bookinfo    reviews-v1       42m   10.10.0.214
bookinfo    reviews-v2       42m   10.10.0.217
bookinfo    reviews-v3       42m   10.10.0.216
```

This is what Kiali currently displays on the workloads tab

![Screenshot from 2021-07-12 20-52-01](https://user-images.githubusercontent.com/6226732/125373393-1475fe80-e353-11eb-945b-b03ddc4e3547.png)

Gathering information for `WorkloadEntry` may vary from the existing workloads since the istio proxy for these workloads run outside of kubernetes.

`WorkloadEntries` represent as single instance of a workload. As `Deployments` are to `Pods` so `WorkloadGroups` are to `WorkloadEntries` and as such Kiali may want to represent `WorkloadGroups` and not the individual `WorkloadEntries` the same way that deployments are represented as a workload and not the individual pod. Although `WorkloadEntries` can exist independently of `WorkloadGroups` in the same way that pods can exist independently of `Deployments`.

**Comment by lucasponce** [comment]
This issue is interesting.
I guess it triggers the question if we want to change the way we deal with Kubernetes vs Istio.
I mean, same happen to other resources, that may be also interpreted as a "Services" (VirtualServices, ServiceEntries).

Typically "Workloads" refer to the workloads that are inside the cluster, perhaps then it may sense to move the Istio Config between other sections, I don't know, I see pros/cons in both approaches.

**Comment by nrfox** [comment]
> I guess it triggers the question if we want to change the way we deal with Kubernetes vs Istio.
> I mean, same happen to other resources, that may be also interpreted as a "Services" (VirtualServices, ServiceEntries).

Yes thanks for bringing this up. It is something to consider and it may require a larger refactor than just adding workload entries/service entries to the graph and/or the navigation tabs. IMO Kiali should focus on the mesh as a whole instead of being specific to a single kubernetes cluster.  This may even help down the road with displaying multiple clusters. As it is, Kiali is not very helpful for workloads running outside of the kubernetes cluster despite those workloads being inside of the mesh.

> Typically "Workloads" refer to the workloads that are inside the cluster, perhaps then it may sense to move the Istio Config between other sections, I don't know, I see pros/cons in both approaches.

Perhaps the Istio Config section can stay focused on the Istio configuration but the Services tab can be expanded to mean any mesh internal service in Istio's Service Registry and Workloads would be any workload that a mesh internal service is pointing to? Gathering information on workloads may be more limited for workloads running outside of the kubernetes cluster.

**Comment by lucasponce** [comment]
> As it is, Kiali is not very helpful for workloads running outside of the kubernetes cluster despite those workloads being inside of the mesh.

+1, I agree in the what, everything that is part of the mesh, should be potentially visualized in Kiali. In the how is something that probably touches the discussions around deployment models.

I'm inclined to show what the mesh "views", I mean a workload can be external VM, but it's a workload type for Kiali.
Yes, the original design didn't consider this case and this would require a different design to make it nice.

> Services tab can be expanded to mean any mesh internal service in Istio's Service Registry and Workloads would be any workload that a mesh internal service is pointing to?

Yes, one things that we have started to brainstorm is that we need a more clean section about "control planes" entities, as they can be many per cluster (isolated, like a federation scenario, or just in a canary deployment), and these control planes have state. Not only about the status, but the "visibility" of the "dataplane" and destinations available.

Today the "Services" section is also coming from the original design, meaning "Kubernetes Services" but the "Istio Registry" deserves also an update here as well, but also, it needs more thinking as in the same cluster, you may have a new important key to define a criteria "Give me the services for control plane A, instead of B".

It's good comment, I think it's part of the deployment model discussions we're having, as to have a explicit view of the control planes is a first step.

**Comment by jshaughn** [comment]
@nrfox Perhaps this deserves separate issues for the config views and the graph.  With respect to the graph, is there an easy way to identify a workload node as a workloadEntry?

**Comment by nrfox** [comment]
> Yes, one things that we have started to brainstorm is that we need a more clean section about "control planes" entities, as they can be many per cluster (isolated, like a federation scenario, or just in a canary deployment), and these control planes have state. Not only about the status, but the "visibility" of the "dataplane" and destinations available.
> 
> Today the "Services" section is also coming from the original design, meaning "Kubernetes Services" but the "Istio Registry" deserves also an update here as well, but also, it needs more thinking as in the same cluster, you may have a new important key to define a criteria "Give me the services for control plane A, instead of B".

Having multiple controlplanes in the mix will definitely affect the design there but could that be a separate issue? Adding the workload entries/service entries to the graph and config views will require a bit of refactoring but that would need to happen anyways regardless of the deployment model. The metadata (`WorkloadEntry` and `ServiceEntry`) objects are still scoped to a single cluster. Where Kiali gets that information from in a multi deployment model I believe would be the same as the other istio resources scoped that particular cluster.

**Comment by nrfox** [comment]
> @nrfox Perhaps this deserves separate issues for the config views and the graph. With respect to the graph, is there an easy way to identify a workload node as a workloadEntry?

Yes these could be broken up into separate issues. The graph view could be done first and link to the workload entry object on the istio config tab the same way that service entries work today. I'm not so familiar with that part of the graph code but I would think yes since it could be as simple as creating a workload node for every `WorkloadEntry`. That does raise another question though of whether `WorkloadEntry` should be displayed at all or just `WorkloadGroup` since `WorkloadEntry` is analogous to a pod (it's a single instance of a workload), `WorkloadGroup` is analogous to a `Deployment` and Kiali does not display individual pods as workload nodes. Rather kiali displays the deployments. However `WorkloadGroup` seems to still be experimental. Perhaps Kiali could display the `WorkloadGroup` as the workload node if it exists, otherwise display the workload entry?

**Comment by lucasponce** [comment]
I think is good to treat external workloads as we treat ServiceEntry fist.
Today the "Workload" concept within Kiali is close to a Kubernetes workload and open it could trigger other redesign that need to be tackle top/down with more perspective.

So, what would I expect as a user ? Identify traffic to an external workload (VM) in the Graph, and navigate into the WorkloadEntry/WorkloadGroup as details, in a first iteration.

**Comment by jshaughn** [comment]
> So, what would I expect as a user ? Identify traffic to an external workload (VM) in the Graph, and navigate into the WorkloadEntry/WorkloadGroup as details, in a first iteration.

I'm +1 on this approach.  We should represent this in the graph and hopefully be able to link to relevant Istio config, and not necessarily a Service or Workload detail.

**Comment by nrfox** [comment]
Based on the conversation above I'm going to split this issue into separate issues as was suggested to better divide the work that needs to be done.

1. WorkloadEntry workload graph nodes should link to WorkloadEntry Istio config. This is a broken link today. 
2. Distinguish between VM based workloads and pod based workloads on the graph. This is still up for debate and can be decided in that issue.
3. Show VM based workloads under the workload details tab. This is also still up for debate and might require a larger refactor.

**Comment by lucasponce** [comment]
> Show VM based workloads under the workload details tab. This is also still up for debate and might require a larger refactor. - This issue.

If the VMs are part of the kubernetes cluster (i.e. KubeVirt) probably this one is clear, but yes, otherwise it opens an "hybrid" model that can expand to other bigger scopes, so +1 to start with points 1 and 2.

Thanks !
